File: benchmarks/fixtures/ldap_injection.java
import javax.naming.NamingException;
import javax.naming.directory.DirContext;
import javax.naming.directory.SearchControls;
import javax.naming.directory.SearchResult;

final class LdapInjectionFixture {
    SearchResult vulnerable(DirContext ctx, String userBase, String userFilter, String username)
            throws NamingException {
        var controls = new SearchControls();
        var filter = userFilter.replace("{0}", username);
        return getSingleResult(ctx, userBase, filter, controls);
    }

    SearchResult patched(DirContext ctx, String userBase, String userFilter, String username)
            throws NamingException {
        var controls = new SearchControls();
        var filter = userFilter.replace("{0}", escapeLdapFilter(username));
        return getSingleResult(ctx, userBase, filter, controls);
    }

    private SearchResult getSingleResult(
            DirContext ctx, String searchBase, String filter, SearchControls controls)
            throws NamingException {
        return ctx.search(searchBase, filter, controls).next();
    }

    private static String escapeLdapFilter(String value) {
        return value.replace("*", "\\2a");
    }
}

File: benchmarks/fixtures/ldap_injection.py
import ldap


def vulnerable(con, search_base, username):
    filter_str = f"(uid={username})"
    return con.search_s(search_base, ldap.SCOPE_SUBTREE, filter_str, ["uid"])


def patched(con, search_base, username):
    escaped_username = ldap.filter.escape_filter_chars(username)
    filter_str = f"(uid={escaped_username})"
    return con.search_s(search_base, ldap.SCOPE_SUBTREE, filter_str, ["uid"])
